CLN: clean color selection in _matplotlib/style

ivanovmg opened the pull request:
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

There was a comment in the top of the module that the code was a bit too dynamic.
The logic in the function was too complex.

I refactored the code
 - Extract functions
 - Simplify logic
 - Add type annotations

Note: there is one issue with mypy, which I do not know how to handle. Currently I suggested to ignore the error.

ivanovmg committed:
REF: extract functions

ivanovmg committed:
CLN: remove try/except/ZeroDivisionError

ivanovmg committed:
REF: drop unnecesasry if statement

ivanovmg committed:
CLN: simplify logic

ivanovmg committed:
DOC: add short docstrings

ivanovmg committed:
CLN: simplify logic further

ivanovmg committed:
TYP: add type annotations

ivanovmg committed:
REF: more explicitly handle string color

ivanovmg committed:
FIX: fix mpl registry reset

charlesdong1991 commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+    if isinstance(color, Iterable) and len(color) == 0:
+        raise ValueError("Invalid color argument: {color}")
+
+    if isinstance(color, dict):
+        return color
+
+    if isinstance(color, str):
+        if _is_single_color(color):
+            # GH #36972
+            return [color]
+        else:
+            return list(color)
+
+    # ignoring mypy error here
+    # error: Argument 1 to "list" has incompatible type
+    # "Union[Sequence[float], Collection[Union[str, Sequence[float]]]]";
+    # expected "Iterable[Union[str, Sequence[float]]]"  [arg-type]
+    # A this point color may be sequence of floats or series of colors,
+    # all convertible to list
+    return list(color)  # type: ignore [arg-type]
```
emm, this seems to be a bit more complex than the original comprehension, is there a reason to write this way?

```python
list(color)	
if is_list_like(color) and not isinstance(color, dict)	
else color
```

ivanovmg replied:
The original portion of the code was exactly quite dynamic. I decided to make it more explicit. I now understand that I missed one case here (which does not hit in tests). This case is when color is sequence of floats (apparently the internal code calls get_standard_colors only with color being sequence of sequence of floats). I will add this edge case and the corresponding test. Presumably, mypy errors will also be resolved.

ivanovmg replied:
Honestly speaking, I kind of liked the original comprehension. The problem is that it was not compatible with typing.

charlesdong1991 commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+    return list(color)  # type: ignore [arg-type]
+
+
+def _get_colors_from_color_type(color_type: str, num_colors: int) -> Collection[Color]:
```
List should be enough

charlesdong1991 commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+        raise ValueError("color_type must be either 'default' or 'random'")
+
+
+def _get_default_colors(num_colors: int) -> Collection[Color]:
```
List should be enough

charlesdong1991 commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+        colors = [c["color"] for c in list(plt.rcParams["axes.prop_cycle"])]
+    except KeyError:
+        colors = list(plt.rcParams.get("axes.color_cycle", list("bgrcmyk")))
+
```
in the original code, seems there is a check of `colors` in this branch, so does it mean after this refactoring, this won't happen? or `colors` shouldn't be a str in the first place once reaching this part of code?
```python
if isinstance(colors, str):	
     colors = list(colors)
```

ivanovmg replied:
Right, in this particular function colors will not be a string. This check is effectively carried out in ``_get_colors_from_color``.

charlesdong1991 reviewed (changes requested):
thanks @ivanovmg 

emm, maybe @simonjayhawkins can provide some feedbacks on the mypy error here?

simonjayhawkins commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+def _get_colors_from_colormap(
+    colormap: Union[str, "Colormap"],
+    num_colors: int,
+) -> Collection[Color]:
```
can use Collection for arguments to be permissive where appropriate, but return type should be more precise, in this case List.

ivanovmg replied:
OK, will adjust for return types. I remember exactly @simonjayhawkins suggesting to use more generic types. Thus, I used collection.

simonjayhawkins replied:
yep. that's correct for function arguments. return types should be precise.

simonjayhawkins commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+    # expected "Iterable[Union[str, Sequence[float]]]"  [arg-type]
+    # A this point color may be sequence of floats or series of colors,
+    # all convertible to list
+    return list(color)  # type: ignore [arg-type]
```
the ci failure could be fixed with 
```suggestion
    return list(color)  # type: ignore[arg-type]
```
but please avoid using ignore instead.

ivanovmg commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+
+
+def _get_colors(
+    *,
```
Just to enforce keyword-only arguments.

ivanovmg commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -1,81 +1,163 @@
+) -> Union[Dict[str, Color], Collection[Color]]:
+    """Get colors from user input color."""
+    if isinstance(color, Iterable) and len(color) == 0:
+        raise ValueError("Invalid color argument: {color}")
```
I added this check to avoid try/except ZeroDivisionError (in the original code), which would be happening when color == "". I will make it an f-string, I missed that.

ivanovmg committed:
TYP: fix typing in _get_colors_from_color

ivanovmg committed:
CLN: eliminate use of legacy "axes.color_cycle"

It turns out that even if "axes.prop_cycle" is not defined,
then it will be composed of CN colors (hex notation).
However, in the current matplotlib there is no "axes.color_cycle"
property in rcParams at all (not even allowed to setup).

ivanovmg committed:
REF: extract generator function to simplify logic

ivanovmg committed:
TST: add tests for get_standard_colors

ivanovmg committed:
CLN: drop list comprehension for generator expr

ivanovmg committed:
TYP: annotate get_standard_colors

ivanovmg committed:
DEP: add testing dependency (cycler)

ivanovmg commented:
Here is a problem. I want to install Cycler (https://matplotlib.org/cycler/) for testing purposes, to mimic matplotlib behavior on default color selection. When I add it into ``environment.yml``, it is rejected by pre-commit check. Moreover, it seems that this package is never installed anyway, which leads to the failure of the tests requiring Cycler.

ivanovmg committed:
Remove test_style temporary

ivanovmg committed:
BLD: remove cycler from dependencies temporary

ivanovmg committed:
Revert "Remove test_style temporary"

This reverts commit 76f76631b9814987482d9b2fe0150f2527e9d25b.

ivanovmg committed:
REF: import cycler from matplotlib.pyplot

ivanovmg commented:
I have some weird problem with the new test module ``test_style.py``. It works fine in my local virtual environment, but fails in CI (some importing problem).

```
_____________ ERROR collecting pandas/tests/plotting/test_style.py _____________
ImportError while importing test module '/home/vsts/work/1/s/pandas/tests/plotting/test_style.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../miniconda3/envs/pandas-dev/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pandas/tests/plotting/test_style.py:5: in <module>
    from pandas.plotting._matplotlib.style import get_standard_colors
pandas/plotting/_matplotlib/__init__.py:3: in <module>
    from pandas.plotting._matplotlib.boxplot import (
pandas/plotting/_matplotlib/boxplot.py:5: in <module>
    from matplotlib.artist import setp
E   ModuleNotFoundError: No module named 'matplotlib'
```

ivanovmg committed:
TST: mark test skip if no mpl

ivanovmg committed:
Merge branch 'master' into cleanup/matplotlib-style

ivanovmg committed:
Merge branch 'master' into cleanup/matplotlib-style

ivanovmg commented:
All checks are good except Windows py38_np18, where tests related to numba failed. I guess it is irrelevant of my changes.

jreback commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -9,74 +19,156 @@
-    # Append more colors by cycling if there is not enough color.
-    # Extra colors will be ignored by matplotlib if there are more colors
-    # than needed and nothing needs to be done here.
+def _cycle_colors(colors: List[Color], num_colors: int) -> List[Color]:
```
is there a reason we don't just use cycler?

ivanovmg replied:
Good question. Probably we can use it. I will check.

jreback replied:
can you update here

ivanovmg replied:
Oh yes, please check my comment https://github.com/pandas-dev/pandas/pull/37203#issuecomment-713441530

jreback replied:
ok cool. pls make an issue about this / separate followon PR. I *would* actually like to use cycler if we can (it post-dates a lot of this code).

jreback commented:
cc @charlesdong1991

ivanovmg commented:
I checked that we actually can use cycler.
But the problem is that tests must be changed.
Currently it is acceptable to return more colors that ``num_colors``.

When I use cycler I would prefer to keep only required number of colors:

```
def _cycle_colors(colors: List[Color], num_colors: int) -> List[Color]:
    import matplotlib.pyplot as plt

    cycle = plt.cycler(color=colors)
    for _, item in zip(range(num_colors), cycle()):
        yield item["color"]

```

This looks simple to understand, IMHO.
And it seems more reasonable that we return exactly ``num_colors`` colors, without any extra (although extra colors are ignored anyway).
But are we good that I go ahead and change some tests?
Or maybe leave it to another PR?

simonjayhawkins commented on pandas/tests/plotting/test_style.py:
```diff
@@ -0,0 +1,162 @@
+
+from pandas import Series
+
+with suppress(ImportError):
+    from pandas.plotting._matplotlib.style import get_standard_colors
+
+
+@td.skip_if_no_mpl
```
you could just use pytest.importorskip see pandas\tests\plotting\test_converter.py or other test modules testing optional dependencies. e.g. pytables

ivanovmg replied:
Did that. Please check if this is what you suggested.

ivanovmg committed:
REF: use pytest.importorskip

charlesdong1991 commented on pandas/tests/plotting/test_style.py:
```diff
@@ -0,0 +1,157 @@
+            (7, ["b", "g", "r", "y", "b", "g", "r"]),
+        ],
+    )
+    def test_default_colors_named_from_prop_cycle_string(self, num_colors, expected):
```
i _think_ you could combine this one with the test above, and add `cycler_color` to parameterize since those two tests look almost identical?

charlesdong1991 commented on pandas/tests/plotting/test_style.py:
```diff
@@ -0,0 +1,157 @@
+    @pytest.mark.parametrize(
+        "num_colors, expected",
+        [
+            (1, ["r", "g", "b", "k"]),
+            (2, ["r", "g", "b", "k"]),
+            (3, ["r", "g", "b", "k"]),
+            (4, ["r", "g", "b", "k"]),
+            (5, ["r", "g", "b", "k", "r"]),
+            (6, ["r", "g", "b", "k", "r", "g"]),
```
i _think_ it's nicer to just test the case where num_color </>/= color, so 3 parametrizations is enough. no need to change here though ^^

charlesdong1991 reviewed (approved):
sorry for late response, i was fallen ill in the past week. 

@ivanovmg  looks very good to me and original logics seem to be covered and unchanged! nice job!

only two minor suggestions on the tests to make the tests slightly shorter, also okay if not change.

WillAyd commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -9,74 +19,156 @@
+        return [color]
+
+    if _is_floats_color(color):
+        color = cast(Sequence[float], color)
```
This cast seems unnecessary - is there an error this solves?

ivanovmg replied:
Yes, cast is necessary.
At this point ``color`` is of type ``Union[Color, Collection[Color]]``. And for some reason ``_is_floats_color`` check does not filter out ``Collection[Color]``. So, instead of ignoring, I added cast here.

WillAyd replied:
Functions can't narrow types yet in mypy, but regardless this is pretty confusing. Can you try to refactor this to make things clearer?

ivanovmg replied:
I do not know how to make it even clearer.
The logic is the following.

1. Start with color being maybe
 a. a single string color
 b. or multiple string color (like 'rgbk')
 c.  or single float color (0.1, 0.2, 0.3)
 d. or multiple float colors
2. Address option a.
3. Address option b.
4. Address options c and d.

Is the logic confusing or the two cast statements are?

ivanovmg replied:
@WillAyd, I updated the logic a little bit.
1. Address one color (either string or sequence of floats)
2. Address collection of colors

Casting is still there as there are no other options.

Please let me know what you think about this implementation.

WillAyd commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -9,74 +19,156 @@
+        color = cast(Sequence[float], color)
+        return [color]
+
+    color = cast(Collection[Color], color)
```
This cast also seems to be hiding a lot of other issues - does this generate an error when removed? `_is_floats_color` imposes a few more conditions than just Sequence versus Collection

ivanovmg replied:
Removing this cast does result in error:
```
pandas\plotting\_matplotlib\style.py:117: error: Argument 1 to "_gen_list_of_colors_from_iterable" has incompatible type "Union[Union[str, Sequence[float]], Collection[Union[str, Sequence[float]]]]"; expected "Collection[Union[str, Sequence[float]]]"  [arg-type]
```

ivanovmg replied:
It is expected that ``_is_floats_color`` will select only those matching type ``Sequence[float]``.
But probably mypy could not comprehend the logic in ``_is_floats_color``.
So it would consider that at the moment of return statement ``color`` variable can be still of type ``Union[Color, Collection[Color]]``, however, in fact it would be only ``Collection[Color]`` after passing through two preceding if statements.

ivanovmg committed:
REF: extract new method _is_single_color

1. Rename _is_single_color -> _is_single_string_color
2. Extract new method _is_single_color, which checks whether
the color provided is a single color, that can be either string
or a sequence of floats.
3. Allow integers to be a part of float sequence color,
(1, 0.4, 0.2, 0.5), for example.

jreback reviewed (changes requested):
looks good. a couple of comments (aside from @WillAyd and @charlesdong1991 )

jreback commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -9,74 +19,174 @@
 ):
-    import matplotlib.pyplot as plt
-
+    if isinstance(color, dict):
```
can you add a doc-strin ghere

ivanovmg replied:
Added. Green. Please take a look.
@jreback

jreback commented on pandas/plotting/_matplotlib/style.py:
```diff
@@ -9,74 +19,174 @@
+    color_type: str,
+    num_colors: int,
+) -> List[Color]:
+    """Get colors from user input."""
```
can you add a doc-string. 

also the current summary is not very  descriptive, nor is the function name .

ivanovmg replied:
I changed it into ``_derive_colors``. I am not sure if you would consider it a better alternative. I thought of having a more verbose name (like ``_derive_colors_from_cmap_color_and_type``, but that does not seem reasonable to me, although I added the explanation in the docstring.
Anyway derive seems a better verb, implying some logic underneath.

jreback reviewed (changes requested):
just a few minor comments, ping on green.

ivanovmg committed:
DOC: add/update docstrings

jreback commented:
thanks @ivanovmg very nice
